pkg/planner/core/rule_predicate_simplification.go

```diff
@@ -147,6 +190,116 @@ func applyPredicateSimplification(sctx base.PlanContext, predicates []expression
+// It is applied for this pattern: P AND (P1 OR P2 ... OR Pn)
+// Pi is removed if P & Pi is false/empty.
+func updateOrPredicate(ctx base.PlanContext, orPredicateList expression.Expression, scalarPredicatePtr expression.Expression) expression.Expression {
+	_, orPredicateType := findPredicateType(orPredicateList)
+	_, scalarPredicateType := findPredicateType(scalarPredicatePtr)
+	scalarPredicateType = comparisonPred(scalarPredicateType)
+	if orPredicateType != orPredicate || scalarPredicateType != scalarPredicate {
+		return orPredicateList
+	}
+	v := orPredicateList.(*expression.ScalarFunction)
+	firstCondition := v.GetArgs()[0]
+	secondCondition := v.GetArgs()[1]
+	_, firstConditionType := findPredicateType(firstCondition)
+	_, secondConditionType := findPredicateType(secondCondition)
+	emptyFirst := false
+	emptySecond := false
+	if comparisonPred(firstConditionType) == scalarPredicate {
+		emptyFirst = unsatisfiable(ctx, firstCondition, scalarPredicatePtr)
+	} else if firstConditionType == orPredicate {
+		firstCondition = updateOrPredicate(ctx, firstCondition, scalarPredicatePtr)
+	}
+	if comparisonPred(secondConditionType) == scalarPredicate {
+		emptySecond = unsatisfiable(ctx, secondCondition, scalarPredicatePtr)
+	} else if secondConditionType == orPredicate {
+		secondCondition = updateOrPredicate(ctx, secondCondition, scalarPredicatePtr)
+	}
+	emptyFirst = emptyFirst || unsatisfiableExpression(ctx, firstCondition)
+	emptySecond = emptySecond || unsatisfiableExpression(ctx, secondCondition)
+	if emptyFirst && !emptySecond {
+		return secondCondition
+	} else if !emptyFirst && emptySecond {
+		return firstCondition
+	} else if emptyFirst && emptySecond {
+		return &expression.Constant{Value: types.NewIntDatum(0), RetType: types.NewFieldType(mysql.TypeTiny)}
+	}
+	return expression.NewFunctionInternal(ctx.GetExprCtx(), ast.LogicOr, v.RetType, firstCondition, secondCondition)
+}
+
+// pruneEmptyORBranches applies iteratively updateOrPredicate for each pair of OR predicate
+// and another scalar predicate.
+func pruneEmptyORBranches(sctx base.PlanContext, predicates []expression.Expression) {
+	if len(predicates) <= 1 {
+		return
+	}
+	for i := range predicates {
+		for j := i + 1; j < len(predicates); j++ {
+			ithPredicate := predicates[i]
+			jthPredicate := predicates[j]
+			_, iType := findPredicateType(ithPredicate)
+			_, jType := findPredicateType(jthPredicate)
+			iType = comparisonPred(iType)
+			jType = comparisonPred(jType)
+			if iType == scalarPredicate && jType == orPredicate {
+				predicates[j] = updateOrPredicate(sctx, jthPredicate, ithPredicate)
+				sctx.GetSessionVars().StmtCtx.SetSkipPlanCache("OR predicate simplification is triggered")
+			} else if iType == orPredicate && jType == scalarPredicate {
+				predicates[i] = updateOrPredicate(sctx, ithPredicate, jthPredicate)
+				sctx.GetSessionVars().StmtCtx.SetSkipPlanCache("OR predicate simplification is triggered")
+			}
```

Comment:
```suggestion
			if iType == scalarPredicate && jType == orPredicate {
				predicates[j] = updateOrPredicate(sctx, jthPredicate, ithPredicate)
			} else if iType == orPredicate && jType == scalarPredicate {
				predicates[i] = updateOrPredicate(sctx, ithPredicate, jthPredicate)
				
			}
		  sctx.GetSessionVars().StmtCtx.SetSkipPlanCache("OR predicate simplification is triggered")
```

Reply:
These two SetSkipPlanCache statements are identical; they can be moved outside.

Reply:
SetSkipPlanCache is applied in "if" and "else if" blocks and not "else". Moving it outside requires another check and does not help.

Reply:
Okay, I was mistaken here.

tokens used: 313